std-bits/table/src/main/java/org/enso/table/data/index/CrossTabIndex.java

```diff
@@ -0,0 +1,214 @@
+package org.enso.table.data.index;
+
+import java.util.*;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+import org.enso.base.text.TextFoldingStrategy;
+import org.enso.table.aggregations.Aggregator;
+import org.enso.table.data.column.builder.Builder;
+import org.enso.table.data.table.Column;
+import org.enso.table.data.table.Table;
+import org.enso.table.error.TooManyColumnsException;
+import org.enso.table.problems.AggregatedProblems;
+import org.enso.table.util.NameDeduplicator;
+import org.graalvm.polyglot.Context;
+
+public class CrossTabIndex {
+  private static final int MAXIMUM_CROSS_TAB_COLUMN_COUNT = 10000;
+
+  private Column[] xColumns;
+
+  private Column[] yColumns;
+
+  private MultiValueIndex<UnorderedMultiValueKey> combinedIndex;
+
+  private ObjectNumberer<UnorderedMultiValueKey> xKeyNumberer;
+
+  private ObjectNumberer<UnorderedMultiValueKey> yKeyNumberer;
+
+  private UnorderedMultiValueKey[][] grid;
+
+  public CrossTabIndex(Column[] xColumns, Column[] yColumns, int tableSize) {
+    this.xColumns = xColumns;
+    this.yColumns = yColumns;
+
+    // Create combined index
+    Column combinedColumns[] =
+        Stream.concat(Arrays.stream(xColumns), Arrays.stream(yColumns)).toArray(Column[]::new);
+    combinedIndex =
+        MultiValueIndex.makeUnorderedIndex(
+            combinedColumns, tableSize, TextFoldingStrategy.unicodeNormalizedFold);
+
+    // Generate lists of combined keys and subkeys
+    List<UnorderedMultiValueKey> combinedKeys = new ArrayList<>(combinedIndex.keys());
+    List<UnorderedMultiValueKey> xSubKeys = new ArrayList<>(combinedKeys.size());
+    List<UnorderedMultiValueKey> ySubKeys = new ArrayList<>(combinedKeys.size());
```

Comment:
If we'd do the `CombinedKey` type, we'd not need these 3 lists anymore - we'd just have the first one and extracting the subkeys would be O(1).